Add SearchControl component

joshuatf opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds a SearchControl component to allow filtering of results, asynchronous requests, and custom rendering of lists.  This may act as a replacement to the existing `SelectControl` component that has a more opinionated render.

Note that styling, tests, and some props are not quite finalized in this PR and will be updated in a follow-up.

<img width="895" alt="Screen Shot 2022-08-02 at 8 47 37 AM" src="https://user-images.githubusercontent.com/10561050/182418652-0e4e7651-3722-4fce-9365-88096762e8c3.png">


Closes  .

### How to test the changes in this Pull Request:

1. Run `npm run storybook`
2. Go to the "SearchControl" component
3. Test the various examples
4. Consider the latest designs and if this components meets those needs

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshuatf committed:
Add initial SearchControl component

joshuatf committed:
Add async example

joshuatf committed:
Reorganize getItemPropsType type

joshuatf committed:
Create separate MenuItem component

joshuatf committed:
Update items to use value/label pairs

joshuatf committed:
Add fuzzy matching example

joshuatf committed:
Use MenuItem component in example

joshuatf committed:
Add callback method example and onSelect prop

joshuatf committed:
Add custom render example

joshuatf committed:
Add styling

joshuatf committed:
Simplify Menu component

joshuatf committed:
Add changelog entry

jacob-sewell reviewed (dismissed):
Storybook examples test well!

A design question: one thing I found myself doing a lot in the Storybook is typing something and then backspacing to see the full list. I'm torn between "that behavior should just happen on focus" and "that behavior could get unmanageable with a large data set, so we should screen out empty search strings".

A couple would-be-nice things I want to flag:

* A README.md file would be :chefs-kiss:.
* I think this component could really benefit from an `emptyText` prop.
* The Gutenberg SearchControl has a magnifying glass icon on the right side of the combobox. Is that one of the style details we'll circle back to, or is it just not part of this version?

joshuatf committed:
Update SelectControl to DeprecatedSelectControl

joshuatf committed:
Rename SearchControl to SelectControl

joshuatf committed:
Add readme

joshuatf committed:
Add placeholder prop

joshuatf committed:
Add icon to combox box

joshuatf committed:
Rename deprecated SelectControl classes

joshuatf committed:
Add changelog entries

joshuatf commented:
Thanks for the review, @jacob-sewell!

>I'm torn between "that behavior should just happen on focus" and "that behavior could get unmanageable with a large data set, so we should screen out empty search strings".

I go back on this behavior too.  In the deprecated component, I previously added `hideBeforeSearch` which allowed us both options.  We could also add this behavior here as well or something more open ended to allow greater control, but we can adjust later if @jarekmorawski has input on what the default behavior should be.

>A README.md file would be :chefs-kiss:.

Added!

>I think this component could really benefit from an emptyText prop.

Added a `placeholder` prop for the input.

>The Gutenberg SearchControl has a magnifying glass icon on the right side of the combobox. Is that one of the style details we'll circle back to, or is it just not part of this version?

I'm unclear on some of the pieces of styling and I guess we'll continue to tweak as we go forward, but I went ahead and added in this icon for the time being.


@ilyasfoo It would be great if you could also give me a sanity check on this since I know you had input on the breaking changes.  I did a major bump for the components and e2e packages and a minor on the woocommerce changes so the former 2 will break without updates and the latter should continue to work, but open to feedback on this.

jarekmorawski commented:
> I'm torn between "that behavior should just happen on focus" and "that behavior could get unmanageable with a large data set, so we should screen out empty search strings".

Great question. People will want to do that, especially in object-based fields like tags or upsells/cross-sells, which rely on existing components. To avoid possible performance issues, maybe we'd add a delay that'd prompt a complete database search while backspacing?

For example, if I typed `shirt` and started removing characters, we wouldn't immediately show search results at `shir`, `shi`, etc. Instead, maybe we'd show the default list with a few recommendations (f.e. recently used items or, to simplify, the first 10), then load the rest after scroll.

ilyasfoo reviewed (commented):
@joshuatf I think the components package version `packages/js/components/package.json` isn't bumped yet? 🤔 

Changelog looks good 👍

joshuatf commented:
@ilyasfoo I thought our changelogs (which specify the type of update) determine the version bump and automatically handle that when we're ready to publish.

Are you thinking we should handle the version bump in this same PR?

jacob-sewell reviewed (commented):
These are good changes! There were a couple of things I noticed this time that are either new or I missed them before:

When you enter a search string that doesn't match anything, you can see the border of the empty menu.
![Screen Shot 2022-08-11 at 16 03 50](https://user-images.githubusercontent.com/890809/184241856-44a56c7f-2761-49d4-9f1c-7e137c709195.png)

I'm getting a lot of downshift errors in my console.
![Screen Shot 2022-08-11 at 16 03 04](https://user-images.githubusercontent.com/890809/184241740-ca28302b-7865-4414-9bd7-26c88cb18043.png)

ilyasfoo commented:
>@ilyasfoo I thought our changelogs (which specify the type of update) determine the version bump and automatically handle that when we're ready to publish.

@joshuatf Ah, I was not aware of that. TIL! In that case, LGTM. 👍 

> Are you thinking we should handle the version bump in this same PR?

My bad, definitely shouldn't handle the version bump in this PR!

ilyasfoo commented:
@joshuatf Sorry if this disrupts your work a little, but I've just [published](https://www.npmjs.com/package/@woocommerce/components/v/10.3.0) components `10.3.0` since it contains some minor fixes that's for a component we released earlier so consumers could get the fix without the upcoming breaking change. 

Hope it also gives a cleaner release for the new component though.

joshuatf commented:
Thanks for the heads up, @ilyasfoo!  No problem at all.

joshuatf committed:
Pass menu props to menu to fix ref issues

joshuatf committed:
Update lock file

jacob-sewell commented on packages/js/components/src/select-control/menu.scss:
```diff
@@ -2,8 +2,16 @@
+    display: none;
+
+    &.is-open {
+        display: block;
```
Nice.

jacob-sewell reviewed (dismissed):
I appreciate freedom the ghost of an empty menu, and the console is also much happier now.

joshuatf committed:
Rebase and update lock file

jacob-sewell commented on pnpm-lock.yaml:
```diff
@@ -347,10 +347,7 @@ importers:
-<<<<<<< HEAD
       '@types/wordpress__media-utils': 3.0.0_react-dom@16.14.0+react@17.0.2
-=======
->>>>>>> 7ec8cece56 (Update lock file)
```
I think removing these lines was a good call.

jacob-sewell reviewed (dismissed):
Still looks good.

joshuatf committed:
Fix up IDs in e2e tests

jacob-sewell reviewed (dismissed):
🚀 🎸

louwie17 commented on packages/js/components/src/select-control/menu.scss:
```diff
@@ -0,0 +1,17 @@
+    width: 100%;
+    top: 100%;
+    box-sizing: border-box;
+    display: none;
```
We should add a `z-index: 10` here as a checkbox shows on top of the list.
Also add a margin-top: $gap-smaller; (is 8px) as the gap seem larger at the moment.

joshuatf replied:
Is this in relation to the categories?  I think we should save that for a follow-up PR.  There's still some styling that needs to be tweaked, but I'd like to see how we plan to use for consumption and make some decisions around the component exports before doing it in the base PR (here).

louwie17 replied:
Yeah when I don't have the `z-index` the checkbox shows up above the dropdown items. But I think in general a z-index makes sense (the last select-control also had this).

louwie17 commented on packages/js/components/src/select-control/menu.tsx:
```diff
@@ -0,0 +1,34 @@
+		>
+			{ isOpen &&
+				( ! Array.isArray( children ) || !! children.length ) && (
+					<div className="woocommerce-select-control__menu-inner">
```
We should change this either to a `li` item or change the parent to a `div` also to stay consistent.

joshuatf replied:
Updated this to be more semantic.

louwie17 commented on packages/js/components/src/select-control/select-control.tsx:
```diff
@@ -0,0 +1,162 @@
+		highlightedIndex,
+		getItemProps,
+		getMenuProps,
+		isOpen,
```
Can we also expose the `getMenuClassname` and `getItemClassname` maybe through the `getProps` functions for each.
It be nice to expose those instead of having to manually copy the existing css names just to match the styling when a user passes their own children in.

joshuatf replied:
I think the point of a consumer using their own is to not inherit styles.  If they want these styles, they should import the necessary components (Menu/MenuItem/etc) to use those styles.

louwie17 replied:
Hmm ok, I was not aware that those were available and exported as well.
It would be worth to add an example of using those in the README as well, this is where I was mostly looking (could be in a separate PR of-course)

louwie17 commented on packages/js/components/src/select-control/select-control.tsx:
```diff
@@ -0,0 +1,162 @@
+	onRemove = () => null,
+	onSelect = () => null,
+	placeholder,
+	selected,
```
Could we also have an option where we can open on focus?

joshuatf replied:
Are you looking for an `onFocus` prop or a way to open the menu when no search query has occurred?  In both cases, I would save for a follow-up.

louwie17 replied:
Depends I guess, if I have the ability to open the menu up outside of the `SelectControl` context then a `onFocus` callback would be great.
If the `isOpen` handling will stay within `SelectControl`, an option for `openOnFocus` would be helpful.

joshuatf replied:
If you override the render with children, you will get access to `isOpen`, but I'm still considering something like an `openOnFocus`.  I was also thinking of introducing this as both an `onFocus` callback and `openMenu`, but still a little on the fence with this.

I'm hoping to see how we use this component in action before settling on one of these paths.

louwie17 commented on packages/js/components/src/select-control/types.ts:
```diff
@@ -0,0 +1,47 @@
+export type ItemType = {
+	value: string;
+	label: string;
+};
```
Could we expose this type as part of the `@woocommerce/components` package. `ItemType` would be to generic, so maybe something like `SelectControlItem`?

Also it would be nice to have an extra optional param with extra data (this is not a must). But in our instance where we are displaying a tree like structure, it be nice to add the children as part of this data.
But I can also work around this, by keeping a key/value store around and making use of this in my own component.

joshuatf replied:
We can expose this in the follow-up.  I'm not happy with exposing the type as it is currently since it doesn't have to follow this format at all.

louwie17 replied:
Ooh really, I thought it had to follow this format in order for the selected items to show up correctly?
If any type can be passed in, we should add support for a generic type.

joshuatf replied:
Technically, any type can be passed in since you can pass your own `itemToString` method.

But agreed, it would be good to have a type for this that's exported.  I'd still love to handle this in a follow-up if possible to unblock other work on this PR, but if you feel strongly about implementing it first here, we can.

louwie17 reviewed (changes requested):
Heya @joshuatf, overall nice work on this, it's working pretty well.
As I am working on the category field dropdown I ran across several small blockers where I made fixes on my end. Given this PR is still open I figured I would make those same suggestions here for now.
There might be more as I continue to work on the category field dropdown, but I can always merge some of that stuff as part of those changes.

joshuatf committed:
Make list structure more semantic

joshuatf commented:
Hey @louwie17 I addressed some of your feedback, but I think a lot of this is pushing the scope of this PR and I'd rather get this merged and continue to iterate than continue to block work for the cycle.

louwie17 committed:
Fix update conflict with pnpm-lock

joshuatf committed:
Move new SelectControl to experimental

joshuatf committed:
Change experimental class name to avoid style conflicts

joshuatf committed:
Fix up latest lock file from rebase

joshuatf committed:
Remove onboarding e2e changes

joshuatf committed:
Update changelogs

joshuatf committed:
Update lock file again

louwie17 committed:
Update pnpm-lock and fix lint error

louwie17 reviewed (approved):
Thanks for moving the component to an experimental component and the continues rebases to fix the `pnpm` conflicts :)
This is good to merge :shipit:
